### ISSUE TITLE
KeyError: 'server'

### ISSUE THREAD

**Issue by mwarkentin** [cloudpassage/jira_halo_issues_sync#43]
I'm getting close to having this working I think - I've been able to connect to halo and fetch issues, but something is going wrong during reconciliation and I haven't been able to track down where this `keyerror` is coming from. I've got this running in docker locally and am continuing to try to debug but a pointer with more knowledge on this codebase might get me moving faster. :)

```
2019-01-18 20:52:28,680 INFO jlib.logger Getting all Halo issues since 2018-12-31T20:52:28.302236
2019-01-18 20:52:32,497 INFO jlib.logger Active issues: 0 Resolved issues: 234.
2019-01-18 20:52:32,497 INFO jlib.logger Discarding 201 issues outside of time range
2019-01-18 20:52:33,315 INFO jlib.logger Reconciling 33 Halo issues since 2018-12-31T20:52:28.302236
2019-01-18 20:52:33,315 INFO jlib.logger Create times between: 2019-01-04T08:18:34.133932Z and 2019-01-17T16:23:35.976581Z
2019-01-18 20:52:33,315 INFO jlib.logger Last seen times between: 2019-01-04T08:18:34.110Z and 2019-01-17T16:23:35.937Z
2019-01-18 20:52:33,315 INFO jlib.logger Resolved times between: 2019-01-17T17:23:36.481Z and 2019-01-17T17:31:57.504Z
2019-01-18 20:52:33,315 INFO jlib.logger Issue type lids: 0
2019-01-18 20:52:33,315 INFO jlib.logger Issue type csm: 0
2019-01-18 20:52:33,316 INFO jlib.logger Issue type fim: 3
2019-01-18 20:52:33,317 INFO jlib.logger Issue type sva: 30
2019-01-18 20:52:33,317 INFO jlib.logger Issue type sam: 0
2019-01-18 20:52:33,317 INFO jlib.logger Issue type fw: 0
2019-01-18 20:52:33,317 INFO jlib.logger Issue type agent: 0
2019-01-18 20:52:33,317 INFO jlib.logger Getting Jira issues which correspond to Halo issues.
2019-01-18 20:52:57,118 INFO jlib.logger Creating and closing Jira issue for Halo issue ID 3d548fec1a7411e9b95f2da22149b99c

Traceback (most recent call last):
  File "/src/jira_halo_issues_sync/application.py", line 70, in <module>
    main()
  File "/src/jira_halo_issues_sync/application.py", line 42, in main
    jlib.Reconciler.reconcile_marching_orders(config, marching_orders)
  File "/src/jira_halo_issues_sync/jlib/reconciler.py", line 46, in reconcile_marching_orders
    pool.map(reconciler_helper, this_batch)
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 251, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 567, in get
    raise self._value
KeyError: 'server'
```

**Comment by mwarkentin** [comment]
From within an ipdb session I've been able to connect to the jira API as well so that doesn't seem to be the issue.

**Comment by mwarkentin** [comment]
I tracked the `KeyError: 'server'` down to the following in my field mapping: `asset_type:halo_asset_type`

Not sure if I misunderstood the docs on how that one should be set.

**Comment by ashmastaflash** [comment]
Field mapping requires the fields exist already in Jira, and that the user account is running as must be able to populate those fields on ticket creation. This is in the screen configuration portion of Jira's configuration. This is the issue you were running into when you got the error: `Field 'halo_issue_id' cannot be set. It is not on the appropriate screen, or unknown.`

Start with no mappings and getting the one required field aligned, which is the `halo_issue_id` field. Once that's in place, adjust your ticket creation screens to accommodate whatever other metadata you want to map. 

Keep in mind that you won't find anything in the mapping that you don't already have in the ticket description. Mappings are there if you need to add more structure (for reporting purposes, for instance) but won't give you any more information than you already have.

**Comment by mwarkentin** [comment]
Thanks! It turned out to be a configuration issue on Jiras side where the project had the custom fields but they weren’t exposed on the create screen which caused issues when creating via api. 

We’ve updated the config and I’ll test on Monday.

**Comment by mwarkentin** [comment]
Got this working, looks like we have to use the `customfield_14124` rather than our nice readable field names.. good ol' jira. 😄 

Going to work on getting the full mapping working and will update if everything is looking good.

**Comment by mwarkentin** [comment]
@ashmastaflash So I'm down the jira rabbithole right now. Here are the things I've tried for halo issue id:

### Set to custom field name (`halo_issue_id`)

* Can search for the issue successfully in `get_jira_issues_for_halo_issue`
* Can't update the field after creation for new issues due to "It is not on the appropriate screen, or unknown." I found [this issue](https://jira.atlassian.com/browse/JRASERVER-62343) for Jira server but couldn't access the Jira Cloud issue to see if it's been resolved..

### Set to custom field ID (`customfield_14124`)

* Can update the field after creation for new issues
* Can't search for the issue in `get_jira_issues_for_halo_issue`: `Field 'customfield_14124' does not exist or you do not have permission to view it.` This is the JQL it's trying to use: `customfield_14124 ~ 3d548fec1a7411e9b95f2da22149b99c`
* It looks like to search by custom field ID there's some special JQL syntax - this works: `cf[14124] ~ 3d548fec1a7411e9b95f2da22149b99c`

I'm trying to figure out if there's a way to make the former (using the custom field name) work as that seems best, but if not it seems like this would require some possible changes to the tool to support the custom field id correctly.

I'm a little confused by how this has been working for you previously - are you on jira server or cloud? Are you using custom fields or built-ins?

Thanks again!

**Comment by mwarkentin** [comment]
I've opened a support ticket w/ atlassian to try to get some insight from their side as well.

**Comment by ashmastaflash** [comment]
In my testing, I'm using the human-friendly field name: https://github.com/cloudpassage/jira_halo_issues_sync/blob/master/.travis.yml#L81

**Comment by mwarkentin** [comment]
@ashmastaflash are you using self-hosted jira or jira cloud?

**Comment by ashmastaflash** [comment]
... though it has also been tested against a self-hosted Jira instance.

**Comment by mwarkentin** [comment]
Ok, good to know - just trying to narrow down what might be causing the difference.

**Comment by mwarkentin** [comment]
First response from support:

```
Ricardo Gebhardt commented:

Hi Michael,

I can confirm from my tests that we are only able to to update a custom field of an issue from a REST call using the custom field's ID, not name.

I will follow up on what happened with JRACLOUD-62343, as it is unavailable for me too.
```

**Comment by ashmastaflash** [comment]
That's interesting. I must have picked a default field in my configuration.

**Comment by mwarkentin** [comment]
Let me know if there's anything I can do to help try to figure this out. :)

**Comment by ashmastaflash** [comment]
Hi @mwarkentin under the covers, this uses the Jira abstraction for Python (https://jira.readthedocs.io/en/master/examples.html#issues). 

We interact with the Jira API to create issues here: https://github.com/cloudpassage/jira_halo_issues_sync/blob/master/jlib/jira_local.py#L40 .  An important thing to note is that we create the issue first (with just the summary, description, issue type, and project), then update it with the mapped metadata fields (line 59). This is where we set the server ID in the ticket.

While writing this tool, I found that it's much easier to use an empty Jira environment for testing and aligning fields. It's very easy to get field sprawl in Jira, so starting small and expanding mapped metadata is definitely the recommended path. I recommend just starting with the very basic configuration (see https://github.com/cloudpassage/jira_halo_issues_sync/blob/master/.travis.yml#L29 for how to do this with Docker).
